### PR TITLE
T2D-210 prepublish access to objects

### PR DESCRIPTION
**Pull request by selvaebi**

### CODE REVIEW

File: src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/Study.java [hunk 1]
```diff
@@ -92,6 +103,26 @@
...
+
+    @PrePersist
+    @PreUpdate
+    public void setCreatedBy() {
+       String userName = LoginUser.getUserName();
+        if ( this.createdBy != null &&  !this.createdBy.equals(userName) ) {
+            throw new AccessDeniedException("Only owners are authorised to Update");
+        }
+        if(this.createdBy == null){
+            this.createdBy = userName;
```
Comment (Zhicheng-Liu) [inline src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/Study.java]: If what I think in my above comment is correct then this is problematic.

File: src/main/java/uk/ac/ebi/ampt2d/metadata/aop/StudyReleaseDateAspect.java [hunk 2]
```diff
@@ -47,9 +49,13 @@ public Object filterOnReleaseDateAdviceFindAll(ProceedingJoinPoint proceedingJoi
...
 
-        QStudy study = QStudy.study;
-        Predicate dateRestrictedPredicate = study.releaseDate.between(null, LocalDate.now()).and(predicate);
+        if (LoginUser.isAdmin()) {
+            return proceedingJoinPoint.proceed(args);
+        }
 
+        QStudy study = QStudy.study;
+        Predicate dateRestrictedPredicate = (study.releaseDate.between(null, LocalDate.now())
+                .or(isUserTheOwnerOrHasPermissionToViewTheObject(study))).and(predicate);
```
[the remark this answers is not in the document]
Comment (selvaebi) [reply src/main/java/uk/ac/ebi/ampt2d/metadata/aop/StudyReleaseDateAspect.java]: Yes i thought of the same , but not loading the aspect if the user is admin,but i didn't try ordering , will try it

File: src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/Study.java [hunk 3]
```diff
@@ -92,6 +103,26 @@
...
     private List<Publication> publications;
 
+    @Column
+    @JsonIgnore
+    private String createdBy;
+
+    @PrePersist
+    @PreUpdate
+    public void setCreatedBy() {
+       String userName = LoginUser.getUserName();
```
[the remark this answers is not in the document]
Comment (selvaebi) [reply src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/Study.java]: i think the submitter of the study will create the metadata/post the metadata. anyhow will ask @cyenyxe

File: src/main/java/uk/ac/ebi/ampt2d/metadata/aop/StudyReleaseDateAspect.java [hunk 4]
```diff
@@ -47,9 +49,13 @@ public Object filterOnReleaseDateAdviceFindAll(ProceedingJoinPoint proceedingJoi
         Object[] args = proceedingJoinPoint.getArgs();
         Predicate predicate = (Predicate) args[0];
 
-        QStudy study = QStudy.study;
-        Predicate dateRestrictedPredicate = study.releaseDate.between(null, LocalDate.now()).and(predicate);
+        if (LoginUser.isAdmin()) {
```
[the remark this answers is not in the document]
Comment (selvaebi) [reply src/main/java/uk/ac/ebi/ampt2d/metadata/aop/StudyReleaseDateAspect.java]: sure will add it

File: src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/Study.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (cyenyxe) [reply src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/Study.java]: The POST request will most probably be run by an automated system (a submission API on top of the ones we are currently building, a push from a message queue, etc).

**Pull request closed without merging**